### PR TITLE
Update ParserBase to support floats directly

### PR DESCRIPTION
**Pull request by pjfanning**
Related to https://github.com/FasterXML/jackson-core/issues/755

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/core/base/ParserBase.java [hunk 1]
```diff
@@ -735,7 +745,15 @@ public float getFloatValue() throws IOException
             _reportError("Numeric value ("+getText()+") out of range of Java float");
         }
         */
-        return (float) value;
+        if ((_numTypesValid & NR_FLOAT) == 0) {
+            if (_numTypesValid == NR_UNKNOWN) {
+                _parseNumericValue(NR_FLOAT);
+            }
+            if ((_numTypesValid & NR_FLOAT) == 0) {
+                convertNumberToFloat();
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/base/ParserBase.java:753]: Is this called from anywhere? Or is it meant to be called by databind and/or Scala module?

Also:

**Comment by cowtowncoder** [comment]
Hmmmh. I think it makes sense, although I have a suspicion there might be a conflict with some of binary backends since they add `_numberFloat` on their own. So it'd be good to have a look at `jackson-dataformats-binary` to see if that is a problem (masking of fields is I think legal but very confusing and likely to cause fials).

File: src/main/java/com/fasterxml/jackson/core/base/ParserBase.java (hunk 1 above)
Comment (pjfanning) [reply src/main/java/com/fasterxml/jackson/core/base/ParserBase.java:753]: tbh much of the change is just me copy/pasting the NR_DOUBLE bits and having NR_FLOAT bits that match - the `_parseSlowFloat` change is the one that fixes the broken test case - it was still broken till I made that additional change - some of the other changes that were made in this PR prior to that may not be necessary

**Comment by pjfanning** [comment]
@cowtowncoder with https://github.com/FasterXML/jackson-databind/pull/3470 merged, the jackson-databind build will fail - is it possible to merge this to fix it?

**Comment by cowtowncoder** [comment]
@pjfanning I'm working on re-creating test at `jackson-core` and will either move databind test under `failing` or doing something -- am aware that in the meantime tests fail. Also hoping to merge this soon, yes!

File: src/main/java/com/fasterxml/jackson/core/base/ParserBase.java [hunk 2]
```diff
@@ -678,6 +686,9 @@ public NumberType getNumberType() throws IOException
         if ((_numTypesValid & NR_BIGDECIMAL) != 0) {
             return NumberType.BIG_DECIMAL;
         }
+        if ((_numTypesValid & NR_FLOAT) != 0) {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/base/ParserBase.java:689]: I have some concerns wrt backwards-compatibility here but...  I guess there's already dependency wrt calls made that affect type so maybe this is fine.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Ok decided that it makes more sense to merge this soon; I think it looks good and if there are issues wrt other parsers, can figure these out later on.

I think that there are probably remaining coercion cases where we do NOT want to coerce between `float`, `double` and/or `BigDecimal` (since results differ from actual parsing) but I think this is a solid first step.

**Comment by cowtowncoder** [comment]
Ok so 2.14 version seems to work with changes, as expected. Getting failures from master (3.0) probably due to refactoring of code (some codepaths not updated appropriately).

**Comment by cowtowncoder** [comment]
Ah, found it. A method from `ParserBase` to new `JsonParserBase`.

**Pull request merged**